src/relax/transform/to_mixed_precision.cc

```diff
@@ -0,0 +1,297 @@
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+/*!
+ * \file src/relax/transform/to_mixed_precision.cc
+ * \brief Automatic mixed precision pass.
+ */
+
+#include <tvm/relax/expr_functor.h>
+#include <tvm/relax/op_attr_types.h>
+#include <tvm/relax/transform.h>
+
+#include "../op/make_op.h"
+
+namespace tvm {
+namespace relax {
+
+enum MixedTypePolicy : int {
+  MIXED_PRECISION_ALWAYS = 0,
+  MIXED_PRECISION_FOLLOW = 1,
+  MIXED_PRECISION_NEVER = 2
+};
+
+// Return array is of type : [MixedTypePolicy (int), bool, Call]
+// Call is a call node, out_dtype_str is the expected output_dtype string
+using FMixedPrecision = runtime::TypedPackedFunc<Array<ObjectRef>(
+    const Call& call_node, const std::string& out_dtype_str)>;
+
+class ToMixedPrecisionMutator : public ExprMutator {
+ public:
+  explicit ToMixedPrecisionMutator(DLDataType output_dtype) : output_dtype_(output_dtype) {}
+
+  void InitVarMap(const Function& func) {
+    for (const auto& param : func->params) {
+      if (const auto* type = param->checked_type_.as<DynTensorTypeNode>()) {
+        UpdateVarMap(param, type->dtype, param);
+      }
+    }
+  }
+
+  // Visit the use-site of a defined Var
+  Expr VisitExpr_(const VarNode* op) {
+    if (const auto* type = op->checked_type_.as<DynTensorTypeNode>()) {
+      auto it = var_map_.find(GetRef<Var>(op));
+      ICHECK(it != var_map_.end());
+      auto type_it = it->second.find(runtime::DLDataType2String(type->dtype));
+      if (type_it != it->second.end()) {
+        return type_it->second;
+      } else {
+        Var cur_var = it->second.begin()->second;
+        return builder_->Emit(MakeCast(cur_var, type->dtype));
+      }
+    } else {
+      return ExprMutator::VisitExpr_(op);
+    }
+  }
+
+  // Visit the use-site of a defined DataflowVar
+  Expr VisitExpr_(const DataflowVarNode* op) {
+    if (const auto* type = op->checked_type_.as<DynTensorTypeNode>()) {
+      auto it = var_map_.find(GetRef<Var>(op));
+      ICHECK(it != var_map_.end());
+      auto type_it = it->second.find(runtime::DLDataType2String(type->dtype));
+      if (type_it != it->second.end()) {
+        return type_it->second;
+      } else {
+        Var cur_var = it->second.begin()->second;
+        return builder_->Emit(MakeCast(cur_var, type->dtype));
+      }
+    } else {
+      return ExprMutator::VisitExpr_(op);
+    }
+  }
+
+  void VisitBinding_(const VarBindingNode* binding) override {
+    auto emit = [this, binding](const Expr& e, const Var& v) -> Var {
```

Comment:
Use builder_->ReEmitBinding, see updated builder infra

Reply:
I need to get the var of the new binding, but ReEmitBinding returns nothing.

Reply:
How about we change ReEmitBinding to return the emitted binding

Reply:
I can try